plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteLikeAndEscapeWithCaseSensitivity.java

```diff
@@ -0,0 +1,99 @@
+{
+    private static final Capture<ConnectorExpression> LIKE_VALUE = newCapture();
+    private static final Capture<ConnectorExpression> LIKE_PATTERN = newCapture();
+    private static final Pattern<Call> PATTERN = call()
+            .with(functionName().equalTo(LIKE_FUNCTION_NAME))
+            .with(type().equalTo(BOOLEAN))
+            .with(argumentCount().matching(count -> count == 2 || count == 3))
+            .with(argument(0).matching(expression().capturedAs(LIKE_VALUE).with(type().matching(VarcharType.class::isInstance))))
+            .with(argument(1).matching(expression().capturedAs(LIKE_PATTERN).with(type().matching(VarcharType.class::isInstance))));
+    // TODO add capture on ESCAPE argument and type
+
+    @Override
+    public Pattern<Call> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<ParameterizedExpression> rewrite(Call expression, Captures captures, RewriteContext<ParameterizedExpression> context)
+    {
+        ConnectorExpression capturedValue = captures.get(LIKE_VALUE);
+        if (capturedValue instanceof Variable variable) {
+            JdbcColumnHandle columnHandle = (JdbcColumnHandle) context.getAssignment(variable.getName());
+            Optional<CaseSensitivity> caseSensitivity = columnHandle.getJdbcTypeHandle().getCaseSensitivity();
+            if (caseSensitivity.orElse(CaseSensitivity.CASE_INSENSITIVE) == CaseSensitivity.CASE_INSENSITIVE) {
+                return Optional.empty();
+            }
+        }
+        Optional<ParameterizedExpression> value = context.defaultRewrite(capturedValue);
+        if (value.isEmpty()) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<QueryParameter> parameters = ImmutableList.builder();
+        parameters.addAll(value.get().parameters());
+        Optional<ParameterizedExpression> pattern = context.defaultRewrite(captures.get(LIKE_PATTERN));
+        if (pattern.isEmpty()) {
+            return Optional.empty();
+        }
+        parameters.addAll(pattern.get().parameters());
+        if (expression.getArguments().size() == 3) {
+            ConnectorExpression escapeExpression = expression.getArguments().get(2);
+            if (!(escapeExpression.getType() instanceof VarcharType)) {
+                return Optional.empty();
+            }
+            Optional<ParameterizedExpression> escape = context.defaultRewrite(escapeExpression);
+            if (escape.isEmpty()) {
+                return Optional.empty();
+            }
+            parameters.addAll(escape.get().parameters());
+            return Optional.of(new ParameterizedExpression(format("%s LIKE %s ESCAPE %s", value.get().expression(), pattern.get().expression(), escape.get().expression()), parameters.build()));
+        }
+
+        return Optional.of(new ParameterizedExpression(format("%s LIKE %s", value.get().expression(), pattern.get().expression()), parameters.build()));
```

Comment:
nit: Can we have a String builder which could capture the `%s LIKE %s` and optionally could add the `ESCAPE` clause towards the end and return a `new ParameterizedExpression`.